Project setup

aksafan opened the pull request:
- Defined project structure.
- Added linters.
- Bumped Express to v5.
- Added security packages and ES6 fixes.
- Added TS support.
- Added sanitize request support to prevent xss attacks.
- Added example action to show accessible functionality:
![image](https://github.com/user-attachments/assets/44319a6f-6925-4328-bfed-a31a2683cc98)
- Added swagger support:
![image](https://github.com/user-attachments/assets/d5547d58-69b6-4022-937f-8f2f9aa5879d)
- Updated Readme according to the new changes.
- Fixed Dependabot [alerts](https://github.com/Code-the-Dream-School/ii-practicum-team-2-back/security/dependabot):
![image](https://github.com/user-attachments/assets/ae9cd267-35aa-48d2-ab23-8720540b554c)
- Added ESM support
- Added bundler support with tsup
- Added tsx to run TS
- Added prettier config

aksafan committed:
Merge branch 'main' into project_setup

aksafan committed:
Update .gitignore

aksafan committed:
Update favicon.ico

aksafan committed:
Add configs

aksafan committed:
Add test folder

aksafan committed:
Add linters

aksafan committed:
Bump express to v5 and add dotenv

aksafan committed:
Add security packages and ES6 fixes

aksafan committed:
Add TS support

aksafan committed:
Add sanitize request support to prevent xss attacks

aksafan committed:
Fix env example port

aksafan committed:
Add service example

aksafan committed:
Add swagger support

IrinaStoetski commented on .gitignore:
```diff
@@ -1,19 +1,48 @@
-logs
-*.log
-npm-debug.log*
+# optional eslint cache
```
.vscode

add plz

aksafan replied:
Done!

IrinaStoetski commented on .prettierrc:
```diff
@@ -0,0 +1 @@
+{}
```
do we need it if it's empty?

IrinaStoetski replied:
I think we'd better define this not to have different settings and then rewrite it because of difference in IDE

aksafan replied:
The idea was to have a boilerplate so we can expand it, with prettier config as well.

If we are good without it, we can remove.
But as you've said there could be a difference in IDE, so better to define it.

Do you have any suggestions what to add here?

aksafan replied:
Never mind, saw your [suggestion](https://github.com/Code-the-Dream-School/ii-practicum-team-2-back/pull/1#discussion_r2009302936), thanks!

IrinaStoetski commented on README.md:
```diff
@@ -1,30 +1,54 @@
-3. Run `npm install` to install dependencies
-4. Pull the latest version of the `main` branch (when needed)
+3. Create `.env` file in the root folder (you can copy from `.env.example` and set it with your vars)
+4. Run `npm install` to install dependencies
```
Run `npm install` to install dependencies inside backend directory

aksafan replied:
Done!

IrinaStoetski commented on README.md:
```diff
@@ -1,30 +1,54 @@
 2. Clone this repository to that folder
-3. Run `npm install` to install dependencies
-4. Pull the latest version of the `main` branch (when needed)
+3. Create `.env` file in the root folder (you can copy from `.env.example` and set it with your vars)
```
is root folder is a backend folder? I think we need to specify

aksafan replied:
Yes, it is. Fixed, thanks!

IrinaStoetski commented on README.md:
```diff
@@ -1,30 +1,54 @@
+
+#### Running ESLint
+
+`npx eslint` - to check everything.
```
why do we use npx instead of npm?

aksafan replied:
Cause we can 😁

Jokes apart, this is from the official [documentation](https://eslint.org/docs/latest/use/getting-started#quick-start) of how to us ESLint with npm:
![image](https://github.com/user-attachments/assets/be6a6e9e-eaab-4d97-98c0-c24a72d1e887)

`npx eslint` runs the local version of eslint installed in our project - a binary from `node_modules/.bin/`.
This is more preferable way as it will work the same across all engineers and on CI/CD.

And we don't need to define something like this:
```
"scripts": {
  "eslint": "eslint ."
}
```
to be able to run `npm run eslint` as we have eslint binary installed out of the box anyways.

IrinaStoetski commented on .prettierrc:
```diff
@@ -0,0 +1 @@
+{}
```
```suggestion
{
  "printWidth": 80,
  "tabWidth": 2,
  "semi": true,
  "singleQuote": false,
  "trailingComma": "es5",
}

```

aksafan replied:
Terrific, thanks a lot!

IrinaStoetski commented on src/controllers/mainController.ts:
```diff
@@ -0,0 +1,36 @@
+  res: Response<CreateUserResponse | ErrorResponse>,
+) => {
+  const { username, email } = req.body;
+  console.log(username);
```
remove

aksafan replied:
It was just to show what we can do with DTOs and types in body and to not let ESLint argue about unused variable.
Removed as you have already seen it.

IrinaStoetski commented on src/middleware/sanitize.ts:
```diff
@@ -0,0 +1,49 @@
+import { NextFunction, Request, Response } from "express-serve-static-core";
+import config from "../config";
+
+function sanitizeValue(value: unknown): unknown {
```
let's use () => everywhere where applicable

aksafan replied:
Agreed, fixed!

IrinaStoetski commented on src/middleware/sanitize.ts:
```diff
@@ -0,0 +1,49 @@
+
+export default function sanitizeRequest(
+  req: Request,
+  _res: Response,
```
what is this logic of using _ in naming?

aksafan replied:
There is a [convention](https://github.com/typescript-eslint/typescript-eslint/issues/1510) to use underscore (`_`) before a parameter name that is required by the function signature but intentionally unused.

IrinaStoetski commented on tsconfig.json:
```diff
@@ -0,0 +1,28 @@
+    "lib": [
+      "ES2020"
+    ],
+    "module": "CommonJS",
```
why not es2015 or esnext
We talked about support of import export

aksafan replied:
It (import/export) is supported. You can observe it in any file of the current branch (e.g. [here](https://github.com/Code-the-Dream-School/ii-practicum-team-2-back/pull/1/files#diff-226c7e3c205fb1d29e63f899a152924e487bf890bb934ecf2542a739504d2a08R5)).

If we change it to `"module": "ESNext",` we will need to add `"type": "module"` into `package.json`.
This will tell Node to treat the project as native ES modules, but Node doesn’t support `.ts` files natively.
Thus, it will require an additional hustle with something like `npx ts-node --loader ts-node/esm src/app.ts` to make it build.

The pros of this will be full ESM compatibility, e.g. top-level await, or ESM-only libraries, that we don't use as far as I know.
The cons will be an additional hustle.

If you believe that we still need full ESM support, we can go for it with additional things.

IrinaStoetski commented on tsconfig.json:
```diff
@@ -0,0 +1,28 @@
+  "compilerOptions": {
+    "target": "ES2020",
+    "lib": [
+      "ES2020"
```
why don't ES2023?

IrinaStoetski replied:
as we work with node 22, what I understood, here is exmple of config for this https://www.npmjs.com/package/@tsconfig/node22 
Found this on typescript website

aksafan replied:
Oh, that's a great catch!

I added [@tsconfig/node-ts](https://npmjs.com/package/@tsconfig/node-ts) and [@tsconfig/node22](https://npmjs.com/package/@tsconfig/node22)

aksafan replied:
Still, I was in need to override these options:
```
"module": "ESNext",
"moduleResolution": "node"
```
To avoid the necessity of always adding `.ts` to all relative imports.

aksafan replied:
Also, I removed both [@tsconfig/node-ts](https://npmjs.com/package/@tsconfig/node-ts) and [@tsconfig/node22](https://npmjs.com/package/@tsconfig/node22) as they can't be extended both at once.
Replaced it with the full tsconfig instead:
```
{
  "compilerOptions": {
    "target": "ES2022",
    "lib": [
      "ES2023"
    ],
    "module": "ESNext",
    "resolveJsonModule": true,
    "rootDir": "./src",
    "moduleResolution": "node",
    "typeRoots": [
      "./node_modules/@types",
      "./src/types"
    ],
    "outDir": "./dist",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true,
    "allowSyntheticDefaultImports": true,
    "baseUrl": ".",
    "paths": {
      "@/*": [
        "src/*"
      ],
      "#root/*": [
        "./*"
      ]
    }
  },
  "include": [
    "src"
  ],
  "exclude": [
    "node_modules",
    "dist"
  ]
}

```

Full modern features support without adding `.ts` to every relative import.

Also it works with a newly installed bundler `tsup` and supports features like:
Instead of using
```
import { getUser } from '../../../services/userService.js';
```
you can use:
```
import { getUser } from '@/services/userService.js';
```

and
```
import { version } from "#root/package.json";
```
for accessing root folder.

IrinaStoetski commented on src/config/xss.ts:
```diff
@@ -0,0 +1,7 @@
+import { FilterXSS } from "xss";
+
+export default new FilterXSS({
+  whiteList: {}, // no tags allowed
```
do we need to specify this at all?

aksafan replied:
Yes, if we want all scripts to be cleaned in request.

IrinaStoetski commented on src/config/swagger.ts:
```diff
@@ -0,0 +1,29 @@
+      },
+    ],
+  },
+  apis: ["./src/routes/**/*.ts", "./src/schema/*.ts", "./src/dtos/*.ts"],
```
We name dtos ---> dto, it is clear enough that there can be multiple

aksafan replied:
Fixed!

IrinaStoetski commented on src/app.ts:
```diff
@@ -0,0 +1,32 @@
+app.use(express.static("public"));
+app.use(favicon(__dirname + "/public/favicon.ico"));
+
+// extra security middlewares
```
// Security 

clear enough

aksafan replied:
Fixed!

IrinaStoetski commented on package.json:
```diff
@@ -1,21 +1,57 @@
 {
-  "name": "bb-practicum-team-1-back",
+  "name": "goal-tracker",
```
goal-tracker-api ? it isn't a big deal though

aksafan replied:
Fixed!

aksafan committed:
Code review fixes

- Added ESM support
- Added bundler support with tsup
- Added tsx to run TS
- Added prettier config

aksafan commented:
@IrinaStoetski thank you for review!

I [fixed](https://github.com/Code-the-Dream-School/ii-practicum-team-2-back/pull/1/commits/2ae546c6bbaea65f52799a799f7de0f2fa88de10) most of issues and answered rest of comments.

IrinaStoetski commented on README.md:
```diff
@@ -1,30 +1,54 @@
+
+#### Running Prettier
+
+`npx prettier . --write` - to format everything.
```
we usually define npm script like "npm run prettier:write" for that. and for lint as well

aksafan replied:
Fixed!

aksafan committed:
Update linter and prettier run commands
